pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/TableQueryQuotaManager.java

```diff
@@ -0,0 +1,265 @@
+    LOGGER.info("Dropping rate limiter for table {}", rawTableName);
+
+    removeRateLimiter(brokerResource, tableName, tableConfig.getQuotaConfig());
+  }
+
+  /** Remove or update rate limiter if another table with the same raw table name but different type is still using the quota config.
+   * @param brokerResource broker resource which stores all the broker states of each table.
+   * @param tableName original table name
+   * @param quotaConfig old quota config
+   * */
+  private void removeRateLimiter(ExternalView brokerResource, String tableName, QuotaConfig quotaConfig) {
+    if (quotaConfig == null || Strings.isNullOrEmpty(quotaConfig.getMaxQueriesPerSecond()) || !_rateLimiterMap.containsKey(tableName)) {
+      // Do nothing if the current table doesn't specify qps quota.
+      LOGGER.info("No qps quota is specified for table: {}", tableName);
+      return;
+    }
+    if (brokerResource == null) {
+      LOGGER.warn("Failed to drop qps quota for table {}. No broker resource connected!", tableName);
+      return;
+    }
+    _rateLimiterMap.remove(tableName);
+  }
+
+  /**
+   * Get QuotaConfig from property store.
+   * @param rawTableName table name without table type.
+   * @param tableType table type: offline or real-time.
+   * @return QuotaConfig, which could be null.
+   * */
+  private QuotaConfig getQuotaConfigFromPropertyStore(String rawTableName, CommonConstants.Helix.TableType tableType) {
+    ZkHelixPropertyStore<ZNRecord> propertyStore = _helixManager.getHelixPropertyStore();
+
+    String tableNameWithType = TableNameBuilder.forType(tableType).tableNameWithType(rawTableName);
+    TableConfig tableConfig = ZKMetadataProvider.getTableConfig(propertyStore, tableNameWithType);
+    if (tableConfig == null) {
+      return null;
+    }
+    return tableConfig.getQuotaConfig();
+  }
+
+  /**
+   * Create a rate limiter for a table.
+   * @param tableName table name with table type.
+   * @param brokerResource broker resource which stores all the broker states of each table.
+   * @param quotaConfig quota config of the table.
+   * */
+  private void createRateLimiter(String tableName, ExternalView brokerResource, QuotaConfig quotaConfig) {
+    if (quotaConfig == null || Strings.isNullOrEmpty(quotaConfig.getMaxQueriesPerSecond())) {
+      LOGGER.info("No qps config specified for table: {}", tableName);
+      return;
+    }
+
+    if (brokerResource == null) {
+      LOGGER.warn("Failed to init qps quota for table {}. No broker resource connected!", tableName);
+      return;
+    }
+
+    Map<String, String> stateMap = brokerResource.getStateMap(tableName);
+    if (stateMap == null) {
+      LOGGER.warn("Failed to init qps quota: table {} is not in the broker resource!", tableName);
+      return;
+    }
+    int onlineCount = 0;
+    for (String state : stateMap.values()) {
+      if (state.equals(CommonConstants.Helix.StateModel.SegmentOnlineOfflineStateModel.ONLINE)) {
```

Comment:
Here you want to check that the state is `ONLINE` and that the instance name is not your instance name (which could happen if the external view is stale).